Better face swapping

LamseyMcBrien opened the pull request:
Sierra is enthused by the idea of being able to swap facesets programmatically, and is probably going to want to make more use of this capability. As such, I have reworked it into its own function rather than piggybacking on custom_bust_display_options.

In addition to being a cleaner separation of concerns, this means we can apply the face swaps even when we're only displaying faces. This will be important for consistency in cases where the busts are suppressed due to visible pictures, or if the user has disabled the busts. Sierra is aware that this will also require changes in other places (e.g. actor config, battle busts, harem menu) in order to be fully consistent.

Sierra also requested that I implement two alternate versions of Uyae - one with no boob window after the end of the Zirantia trip, then another with an updated boob window after the start of ch4.

This builds on and supersedes #11 since it'd be a pain with conflicts to merge them separately. You can check f27845bf2dab46d894ad317b202218c12e499d1a to see just the new changes.

Rachnera commented on Bust Config.rb:
```diff
@@ -157,8 +158,49 @@
+    return $game_message.face_name
+  end
+  
   def custom_bust_display_options
```
`custom_bust_display_options` was a mistake and can now be removed (Vera is also a faceswap). Allows to get rid of `bust_display_options` while we're at it.

Rachnera commented on Graphics/Faces/Uyae3 emo.png:
It was also a mistake to allow you to duplicate Uyae's faces the first time...
You should be able to achieve the same endgame (associate different busts to the same faces) without keeping multiplying files as Ivala multiplied breads by hacking [around here](https://github.com/Rachnera/tls-experiments/blob/a8ee8aba006c9c487863f7797050383a767fd188/Rachnera%20Message%20Bust.rb#L98).

Rachnera reviewed (commented):
You know I aren't going to merge the dozen identical Simon's faces right? Those would be a maintenance nightmare we have no reason to inflict on anyone.

Rachnera commented:
Conflict is from me doing the cleanup I was talking about yesterday, as it indeed wasn't very nice to try to push it on someone else:
https://github.com/Rachnera/tls-experiments/commit/d3ef666a81652d2e3c47ce9e6ba1b71d68c9bf93
https://github.com/Rachnera/tls-experiments/commit/9a48c65287f2bd6ff3ee16e846ef120e28eebd46

LamseyMcBrien committed:
Face swapping updates

- Auto face swapping applies to both bust and non-bust dialogue
- Topless Simon support
- Uyae boob window updates requested by SL

LamseyMcBrien commented:
I have confirmed with Sierra that she is OK with not supporting topless young Simon, and having the topless variants listed only next to face002b when editing events. This allows us to reduce down to 4 topless Simon facesets that cover both face002b and 1 Simon.

I have also confirmed with Sierra that she's OK with being unable to override the Uyae2/Uyae3 bust display via faceset selection, which allows the duplicate faceset for Uyae3 to be dropped.

Uyae2 emo is not a duplicate faceset (altered collar and shoulder). Please don't delete it again.

After conflict resolution, I've reinstated the changes from my branch to make faceset swapping apply to non-busty dialogue boxes. As part of this, I have cleaned up the handling of Vera's face in Y DS_Actor 17 since it can now be handled entirely by the auto faceset swap.

I have retested all of the above on a fresh extract of 0.75.5.

Rachnera commented on Rachnera Message Bust.rb:
```diff
@@ -364,7 +364,7 @@ def valid_context?
 
   def character_name
-    Busty::character_from_face($game_message.face_name, $game_message.face_index)
+    Busty::character_from_face(bust_face_name, $game_message.face_index)
```
Oh, you've removed all references to `character_name` from within `bust_face_name` to avoid the infinite loop.

Should work better that way indeed as, unless I'm mistaken, that means the anticrash code in case of missing busts and similar shenanigans should run properly now. 👍 

(probably should replace `$game_message.face_index` with `bust_face_index` for consistency sake; but that's like super nitpicky)

LamseyMcBrien replied:
Yeah, I recall this also being important to make sure that the correct bust config (offsets etc) was selected. Fair point about the index, I'll test that shortly and update.

LamseyMcBrien replied:
Done.

Rachnera commented on Rachnera Message Bust.rb:
```diff
@@ -417,8 +417,8 @@ def bust_extra_x
-    return original_591_draw_face(face_name, face_index, x + 2*standard_padding, y, enabled) if valid_context?
-    original_591_draw_face(face_name, face_index, x, y, enabled)
+    return original_591_draw_face(bust_face_name, face_index, x + 2*standard_padding, y, enabled) if valid_context?
+    original_591_draw_face(bust_face_name, face_index, x, y, enabled)
```
This feature was way past the point where it was just going to be removed. But hacking the default behavior regardless of the switch officializes it in a way.

LamseyMcBrien replied:
Yeah, this is intentional since Sierra is keen on the ability to auto-swap appearances as a means of showing character progression, so I think it makes sense to apply the faceset swaps even if the dialogue busts are hidden. I guess it depends on whether you regard the setting as toggling the display of dialogue busts or disabling the script altogether.

Rachnera commented on Bust Config.rb:
```diff
@@ -157,24 +151,59 @@
+    # if Simon is using his almost-naked sprite, override with topless faceset, unless it's young Simon
+    topless_simon_face = Busty::TOPLESS_SIMON_MAP[$game_message.face_name]
+    if topless_simon_face && $game_actors[2].character_index == 2 && 
+        ($game_message.face_index != 7 || !Busty::SIMON_YOUNG_FACES.include?(topless_simon_face))
```
The condition seems weird to me as, as written, it seems it will fail on the perfectly fine `1 Simon dark eyes` eight face?

Maybe something more straightforward like?
```rb
young_simon = ["face002b", "face002b dark"].include?($game_message.face_name) && $game_message.face_index == 7
```

LamseyMcBrien replied:
That if statement translates to "If we have a topless version of the faceset being shown, and Simon's character sprite is naked, and the face being displayed isn't young Simon, then swap the face", where "the face being displayed isn't young Simon" is true if the face index isn't 7 or the face isn't one of the ones with Young Simon. I guess I could invert that last bit to make it less confusing.

LamseyMcBrien replied:
Done.

Rachnera reviewed (commented):
👍 

Just a second check on the young Simon condition and good for merge.

(sorry for the Uyae stuff; got overzealous there)

LamseyMcBrien committed:
PR review fixes
